### PR TITLE
[WIP] IncrementalPCA in CanICA

### PR DESCRIPTION
**Pull request by lesteve**
The idea is to reduce memory usage to be able to tackle data like HCP.

### CODE REVIEW

**Comment by lesteve** [comment]
IncrementalPCA only exists in scikit-learn 0.16.1. Because of that the only tests we expect to pass at the moment is the python 3.4 ones using the latest versions of scikit-learn.

To do/questions:
- [ ] Failing tests: for some reason giving twice the same data doesn't produce the same eigenvectors in MultiPCA. Not sure how bad this is.
- [ ] Do we want to make this feature available for scikit-learn versions older than 0.16 ? From a quick look at it, it would be slightly painful.
- [ ] How do we want to support this feature, a boolean parameter in CanICA to toggle IncrementalPCA (False by default) ?
- [ ] Caching the result of the svd to mirror the previous behaviour

**Comment by lesteve** [comment]
Also not sure how to evaluate the quality of the maps.

plot_canica_resting_state.py 

### master
canica_plots_master

### this PR
canica_plots_mine

**Comment by lesteve** [comment]
Some memory_profiler plots. Data is from @mrahim (221 subjects from ADNI, ~18GB):
### master
canica_master_memory_profile

### this PR
canica_incremental_pca_memory_profile

Memory usage decreased from 8GB to 2.5GB.
Running time increased by a factor 2.

**Comment by bthirion** [comment]
On 17/06/2015 16:01, Loïc Estève wrote:
>
> Also not sure how to evaluate the quality of the maps.
>
It is hard to say anything on quality, but it is sueful to quantify how 
different the result is between two implementations. For this, I would 
use a hungarian algorithm and report the sums of correlations of matched 
components.

**Comment by GaelVaroquaux** [comment]
Great job! The maps look good. No problem with the variability that you observe.

The memory reduction is great! It would be nice to use a profiler to see where the cost comes in terms of time. I suspect that doing a PR to scikit-learn to use randomized_svd in the IncrementalPCA would give you the speed back.

Finally: we won't try to support this for versions of scikit-learn older than 0.16.

**Comment by lesteve** [comment]
I changed a few things, most notably the fact that there is a `group_incremental_pca` parameter in CanICA to use IncrementalPCA for the group PCA.

As @ogrisel noticed the IncrementalPCA is going to center the data in contrary to randomized_svd, which is what is used in master, so we shouldn't expect matching between the two.

Still haven't figured out why the tests were failing.

File: nilearn/decomposition/multi_pca.py [hunk 1]
```diff
@@ -188,6 +196,20 @@ def __init__(self, n_components=20, smoothing_fwhm=None, mask=None,
         self.target_affine = target_affine
         self.target_shape = target_shape
         self.standardize = standardize
+        self.incremental_group_pca = incremental_group_pca
+
+        if incremental_group_pca:
+            try:
+                from sklearn.decomposition.incremental_pca import IncrementalPCA
+            except ImportError as exc:
```
Comment (banilo) [inline nilearn/decomposition/multi_pca.py]: How about a simple warning instead of an exception in IncrementalPCA should not be available?

**Comment by banilo** [comment]
> Failing tests: for some reason giving twice the same data doesn't produce the same eigenvectors in MultiPCA. Not sure how bad this is.

Given SVD should return the same result every time, is it not more likely that non-identical eigenvectors from MultiPCA are related to a) multi-processing and/or b) caching mechansims. Perhaps debug by ruling out joblib sources...?

**Comment by lesteve** [comment]
> Given SVD should return the same result every time, is it not more likely that non-identical eigenvectors from MultiPCA are related to a) multi-processing and/or b) caching mechansims. Perhaps debug by ruling out joblib sources...?

I am afraid there is a misunderstanding here. The failing test is [here](https://github.com/nilearn/nilearn/blob/master/nilearn/decomposition/tests/test_multi_pca.py#L35):

```python
components1 = multi_pca.fit(data).components_
components2 = multi_pca.fit(2 * data).components_
np.testing.assert_array_almost_equal(components1, components2)
```

`data` is a list of Nifti images. Twice the same data means `2 * data`. Unfortunately I don't have a good enough understanding of CanICA to know how important this test failure is.

**Comment by arthurmensch** [comment]
MultiPCA output different results depending on the run because of a lack of `random_state` in `session_pca`. I have a working branch where this issue is solved, I should clean it before proposing it.

File: nilearn/decomposition/multi_pca.py [hunk 2]
```diff
@@ -360,3 +372,44 @@ def inverse_transform(self, component_signals):
...
+                pca.partial_fit(subject_pca)
+
+            return pca.components_, pca.singular_values_
+
+        return cache(_incremental_group_pca_impl,
+                     self.memory,
+                     func_memory_level=3,
+                     memory_level=self.memory_level)(
+                         subject_pcas, self.n_components)
+
```
Comment (arthurmensch) [inline nilearn/decomposition/multi_pca.py]: Is this the canonical way of hiding cached functions ? This is not what is used in other parts of nilearn. It looks clean though

**Comment by bthirion** [comment]
On 03/07/2015 12:51, Arthur Mensch wrote:
>
> MultiPCA output different results depending on the run because of a 
> lack of |random_state| in |session_pca|. I have a working branch where 
> this issue is solved, I should clean it before proposing it.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nilearn/nilearn/pull/613#issuecomment-118314510>.
>
Would be great indeed.

File: nilearn/decomposition/multi_pca.py [hunk 3]
```diff
@@ -360,3 +372,44 @@ def inverse_transform(self, component_signals):
...
+    def _incremental_group_pca(self, subject_pcas):
+        def _incremental_group_pca_impl(subject_pcas, n_components):
+            # delayed import because IncrementalPCA is only supported for
+            # sklearn versions > 0.16
+            from sklearn.decomposition.incremental_pca import IncrementalPCA
+
+            pca = IncrementalPCA(n_components=self.n_components)
+            for subject_pca in subject_pcas:
+                pca.partial_fit(subject_pca)
```
Comment (arthurmensch) [inline nilearn/decomposition/multi_pca.py]: I think we should work out a way on lazy loading subject pcas : On HCP we typically use 100 components for 1650 records, which approximately reduce the whole data set of a factor 10 (1200 samples per records) : `subject_pcas` will be around 150GB, which is too much for a laptop.

The idea would be to mini-batch subjects pcas so as to benefit from parallelization : we should partial fit on batch of `subject_pcas` of size around `n_jobs`. This would require to put Parallel call into `session_pca`, and transform this function into a generator.

Theoretically, we should then keep in memory `batch_size * single_subject_pca_size * 2`, which should keep under 2GB even on HCP.

Comment (lesteve) [inline nilearn/decomposition/multi_pca.py]: Another approach which is probably easier to implement (proposed by @GaelVaroquaux): use joblib.cache_and_shelve to reduce the memory usage.

Comment (AlexandreAbraham) [inline nilearn/decomposition/multi_pca.py]: > This would require to put Parallel call into session_pca, and transform this function into a generator.

This code already exists and is part of a PR that will be proposed after #585 is merged.

> use joblib.cache_and_shelve to reduce the memory usage.

Not a good idea because of the IO overhead.

File: nilearn/decomposition/multi_pca.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/decomposition/multi_pca.py]: This is the proper way if you are in function. Here, this code could be optimized by inheriting CacheMixin and using self._cache.

**Comment by AlexandreAbraham** [comment]
> MultiPCA output different results depending on the run because of a lack of random_state in session_pca, which critically call randomized_svd when data is too large. I have a working branch where this issue is solved, I should clean it before proposing it.

I think that the problem is that the results are stable when using `randomized_pca` and not stable when using incremental PCA.

File: nilearn/decomposition/multi_pca.py (hunk 3 above)
Comment (GaelVaroquaux) [inline nilearn/decomposition/multi_pca.py]: > I think we should work out a way on lazy loading subject pcas : On HCP we
> typically use 100 components for 1650 records, which approximately reduce the
> whole data set of a factor 10 (1200 samples per records) : subject_pcas will be
> around 150GB, which is too much for a laptop.

The right thing to do is probably to use joblib shelving:
https://pythonhosted.org/joblib/memory.html#shelving-using-references-to-cached-values

Talk to @AlexandreAbraham: he has been using this in his related code.

Comment (GaelVaroquaux) [inline nilearn/decomposition/multi_pca.py]: >     use joblib.cache_and_shelve to reduce the memory usage.

> Not a good idea because of the IO overhead.

Should be optional, with an auto parameter that turns it on if the list
of subjects is very big (eg more than 50).

**Comment by lesteve** [comment]
I rescucitated this PR after the changes that went in related to the DictionaryLearning. Here are the updated maps coming from examples/03_connectivity/plot_canica_resting_state.py:

#### Master
canica_components_master

#### This PR with incremental_group_pca=True
canica_components_pr_incremental_group_pca

There is still the same test failing as previously, not sure what to do about it.

File: nilearn/decomposition/multi_pca.py [hunk 4]
```diff
@@ -164,11 +190,39 @@ def _raw_fit(self, data):
...
+
+
+def _incremental_group_pca(subject_pcas, n_components, nb_subjects):
+    # delayed import because IncrementalPCA is only supported for
+    # sklearn versions > 0.16
+    from sklearn.decomposition.incremental_pca import IncrementalPCA
+
+    pca = IncrementalPCA(n_components=n_components)
+    subject_length = len(subject_pcas) // nb_subjects
+    subject_indices = (slice(i, i + subject_length)
```
Comment (lesteve) [inline nilearn/decomposition/multi_pca.py]: This is a bit hacky because now subject_pcas come in a single array instead of a list of per-subject PCA and the concactenation happens a bit deep in the BaseDecomposition code. I did that with the old code in mind but maybe I don't actually need to bother doing it ...

**Comment by arthurmensch** [comment]
Given that making IO is the bottleneck here, I think this would greatly
benefit from mask shelving.

**Comment by bthirion** [comment]
I don't see any blocker for merging that one.

**Comment by AlexandreAbraham** [comment]
> I don't see any blocker for merging that one.

This one is still not merged because of weird instabilities when using Incremental PCA compared to regular or randomized PCA. If you say this is OK, we can lower the requirements to make the test pass and ignore these results.

**Comment by lesteve** [comment]
There is also this comment: https://github.com/nilearn/nilearn/pull/613#discussion_r61042980. Not sure whether we care how we split the data into batches to call partial_fit on.

**Comment by kchawla-pi** [comment]
Is this alive? If not, please close it by this weekend.
